File: backend/app/models/registry.py
"""Infrastructure Registry models."""
from datetime import datetime
from sqlalchemy import Boolean, Column, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import relationship

from app.core.database import Base


class RegistryDomain(Base):
    __tablename__ = "registry_domains"

    id = Column(String, primary_key=True)
    organization_id = Column(String, ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False)
    server_id = Column(String, ForeignKey("servers.id", ondelete="SET NULL"), nullable=True)
    name = Column(String, nullable=False)
    registrar = Column(String, nullable=True)
    registered_at = Column(DateTime, nullable=True)
    expires_at = Column(DateTime, nullable=True)
    auto_renew = Column(Boolean, nullable=False, default=False)
    nameservers = Column(Text, nullable=True)
    notes = Column(Text, nullable=True)
    tags = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    ssl_certs = relationship("RegistrySslCert", back_populates="domain", cascade="all, delete-orphan")


class RegistrySslCert(Base):
    __tablename__ = "registry_ssl_certs"

    id = Column(String, primary_key=True)
    organization_id = Column(String, ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False)
    domain_id = Column(String, ForeignKey("registry_domains.id", ondelete="SET NULL"), nullable=True)
    domain_name = Column(String, nullable=False)
    common_name = Column(String, nullable=True)
    issuer = Column(String, nullable=True)
    issued_at = Column(DateTime, nullable=True)
    expires_at = Column(DateTime, nullable=True)
    fingerprint = Column(String, nullable=True)
    auto_renew = Column(Boolean, nullable=False, default=False)
    provider = Column(String, nullable=True)
    notes = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    domain = relationship("RegistryDomain", back_populates="ssl_certs")


class RegistryDbService(Base):
    __tablename__ = "registry_db_services"

    id = Column(String, primary_key=True)
    organization_id = Column(String, ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False)
    server_id = Column(String, ForeignKey("servers.id", ondelete="SET NULL"), nullable=True)
    name = Column(String, nullable=False)
    db_type = Column(String, nullable=False, default="postgres")
    host = Column(String, nullable=False)
    port = Column(Integer, nullable=True)
    database_name = Column(String, nullable=True)
    username = Column(String, nullable=True)
    password_enc = Column(Text, nullable=True)
    environment = Column(String, nullable=False, default="prod")
    ssl_enabled = Column(Boolean, nullable=False, default=False)
    notes = Column(Text, nullable=True)
    tags = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)


class RegistryCredential(Base):
    __tablename__ = "registry_credentials"

    id = Column(String, primary_key=True)
    organization_id = Column(String, ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False)
    server_id = Column(String, ForeignKey("servers.id", ondelete="SET NULL"), nullable=True)
    name = Column(String, nullable=False)
    credential_type = Column(String, nullable=False, default="web_login")
    url = Column(String, nullable=True)
    username = Column(String, nullable=True)
    secret_enc = Column(Text, nullable=True)
    environment = Column(String, nullable=False, default="prod")
    notes = Column(Text, nullable=True)
    tags = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)


class RegistryAuditLog(Base):
    __tablename__ = "registry_audit_log"

    id = Column(String, primary_key=True)
    organization_id = Column(String, ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False)
    user_id = Column(String, ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    asset_type = Column(String, nullable=False)
    asset_id = Column(String, nullable=False)
    action = Column(String, nullable=False)
    ip_address = Column(String, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)

File: frontend/src/api/opspilot/registry.ts
import request from './client'

export interface Domain {
  id: string
  org_id: string
  server_id?: string
  name: string
  registrar?: string
  registered_at?: string
  expires_at?: string
  auto_renew: boolean
  nameservers?: string
  notes?: string
  tags?: string
  created_at: string
}

export interface SslCert {
  id: string
  org_id: string
  domain_id?: string
  domain_name: string
  common_name?: string
  issuer?: string
  issued_at?: string
  expires_at?: string
  fingerprint?: string
  auto_renew: boolean
  provider?: string
  notes?: string
  created_at: string
}

export interface DbService {
  id: string
  org_id: string
  server_id?: string
  name: string
  db_type: string
  host: string
  port?: number
  database_name?: string
  username?: string
  has_password: boolean
  environment: string
  ssl_enabled: boolean
  notes?: string
  tags?: string
  created_at: string
}

export interface Credential {
  id: string
  org_id: string
  server_id?: string
  name: string
  credential_type: string
  url?: string
  username?: string
  has_secret: boolean
  environment: string
  notes?: string
  tags?: string
  created_at: string
}

export interface AuditLog {
  id: string
  user_id?: string
  asset_type: string
  asset_id: string
  action: string
  ip_address?: string
  created_at: string
}

export const RegistryAPI = {
  listDomains: (): Promise<Domain[]> => request.get('/registry/domains'),
  createDomain: (data: Record<string, unknown>): Promise<Domain> => request.post('/registry/domains', data),
  updateDomain: (id: string, data: Record<string, unknown>): Promise<Domain> =>
    request.put(`/registry/domains/${id}`, data),
  deleteDomain: (id: string): Promise<void> => request.delete(`/registry/domains/${id}`),

  listSslCerts: (): Promise<SslCert[]> => request.get('/registry/ssl-certs'),
  createSslCert: (data: Record<string, unknown>): Promise<SslCert> => request.post('/registry/ssl-certs', data),
  updateSslCert: (id: string, data: Record<string, unknown>): Promise<SslCert> =>
    request.put(`/registry/ssl-certs/${id}`, data),
  deleteSslCert: (id: string): Promise<void> => request.delete(`/registry/ssl-certs/${id}`),

  listDatabases: (): Promise<DbService[]> => request.get('/registry/databases'),
  createDatabase: (data: Record<string, unknown>): Promise<DbService> => request.post('/registry/databases', data),
  updateDatabase: (id: string, data: Record<string, unknown>): Promise<DbService> =>
    request.put(`/registry/databases/${id}`, data),
  deleteDatabase: (id: string): Promise<void> => request.delete(`/registry/databases/${id}`),
  revealDbPassword: (id: string): Promise<{ value: string }> => request.get(`/registry/databases/${id}/reveal`),

  listCredentials: (): Promise<Credential[]> => request.get('/registry/credentials'),
  createCredential: (data: Record<string, unknown>): Promise<Credential> => request.post('/registry/credentials', data),
  updateCredential: (id: string, data: Record<string, unknown>): Promise<Credential> =>
    request.put(`/registry/credentials/${id}`, data),
  deleteCredential: (id: string): Promise<void> => request.delete(`/registry/credentials/${id}`),
  revealCredentialSecret: (id: string): Promise<{ value: string }> => request.get(`/registry/credentials/${id}/reveal`),

  getAuditLog: (asset_id?: string): Promise<AuditLog[]> =>
    request.get('/registry/audit-log', { params: asset_id ? { asset_id } : {} }),
  search: (q: string): Promise<{ type: string; id: string; name: string; subtitle: string }[]> =>
    request.get('/registry/search', { params: { q } }),
}
